packages/angular/ssr/src/request.ts

```diff
@@ -0,0 +1,74 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import type { IncomingHttpHeaders, IncomingMessage } from 'node:http';
+
+/**
+ * Converts a Node.js `IncomingMessage` into a Web Standard `Request`.
+ *
+ * @param nodeRequest - The Node.js `IncomingMessage` object to convert.
+ * @returns A Web Standard `Request` object.
+ */
+export function createWebRequestFromNodeRequest(nodeRequest: IncomingMessage): Request {
+  const { headers, method = 'GET' } = nodeRequest;
+  const withBody = method !== 'GET' && method !== 'HEAD';
+
+  return new Request(createRequestUrl(nodeRequest), {
+    method,
+    headers: createRequestHeaders(headers),
+    body: withBody ? nodeRequest : undefined,
+    duplex: withBody ? 'half' : undefined,
+  });
+}
+
+/**
+ * Creates a `Headers` object from Node.js `IncomingHttpHeaders`.
+ *
+ * @param nodeHeaders - The Node.js `IncomingHttpHeaders` object to convert.
+ * @returns A `Headers` object containing the converted headers.
+ */
+function createRequestHeaders(nodeHeaders: IncomingHttpHeaders): Headers {
+  const headers = new Headers();
+
+  for (const [name, value] of Object.entries(nodeHeaders)) {
+    if (typeof value === 'string') {
+      headers.append(name, value);
+    } else if (Array.isArray(value)) {
+      for (const item of value) {
+        headers.append(name, item);
+      }
+    }
+  }
+
+  return headers;
+}
+
+/**
+ * Creates a `URL` object from a Node.js `IncomingMessage`, taking into account the protocol, host, and port.
+ *
+ * @param nodeRequest - The Node.js `IncomingMessage` object to extract URL information from.
+ * @returns A `URL` object representing the request URL.
+ */
+function createRequestUrl(nodeRequest: IncomingMessage): URL {
```

Comment:
Question: Was there any specific documentation you were following here? There seems to be a lot of edge cases in `IncomingMessage` behavior and checking multiple locations for each value worries me that we're not checking all of them. The `headers[':authority']` in seems particularly unintuitive and unexpected. It feels like there could easily be some other case where we have to check something non-obvious like `headers[':address-containing-the-port-in-some-format']` to extract port information, and hard to prove that isn't ever necessary.

How can we have confidence that this function is properly covering all the relevant edge case behavior of `IncomingMessage` beyond just trial and error bug fixing?

Reply:
There is no specific documentation on this. It is a combination of behavior observed in other frameworks, such as Nitro and Astro.

Reply:
That's unfortunate as this seems like a fairly complicated API to use. I can't think of anything we can really do to mitigate though beyond the tests you already have. We might want to spend some extra time testing with different Node server frameworks or solicit feedback during developer preview to see if there are edge cases we might be missing here.